kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/CsrServiceImpl.java

```diff
@@ -156,31 +156,62 @@ private boolean isMsisdnInvalid(CallSummaryRecord record) {
     }
 
 
+    private boolean isFirstMessageOfWeek(String weekId) {
+        return weekId.endsWith("1");
+    }
+
+
     private void rescheduleCall(Subscription subscription, CallSummaryRecord record, CallRetry callRetry) {
 
         Long msisdn = subscription.getSubscriber().getCallingNumber();
 
-        if (callRetry == null) {
-
-            // This message was never retried before, so this is a first retry
-            LOGGER.debug(String.format("Creating retry for subscription: %s", subscription.getSubscriptionId()));
-            DayOfTheWeek nextDay = DayOfTheWeek.fromInt(subscription.getStartDate().dayOfWeek().get()).nextDay();
-            CallRetry newCallRetry = new CallRetry(
-                    subscription.getSubscriptionId(),
-                    msisdn,
-                    nextDay,
-                    CallStage.RETRY_1,
-                    record.getContentFileName(),
-                    record.getWeekId(),
-                    getLanguageCode(subscription),
-                    getCircleName(subscription),
-                    subscription.getOrigin()
-            );
-            callRetryDataService.create(newCallRetry);
+        if (callRetry == null || !callRetry.getWeekId().equals(record.getWeekId())) {
```

Comment:
This might be bordering on a nitpick but this might be a cleaner flow -

        if (callRetry == null) {
            // no retry was ever tried

            CallRetry newCallRetry = new CallRetry(
                    subscription.getSubscriptionId(),
                    msisdn,
                    null,
                    CallStage.RETRY_1,
                    record.getContentFileName(),
                    record.getWeekId(),
                    getLanguageCode(subscription),
                    getCircleName(subscription),
                    subscription.getOrigin()
            );
            callRetryDataService.create(newCallRetry);
            return;
        }

        if (!callRetry.getWeekId().equals(record.getWeekId())) {

            // this call was retried for a different week but never removed from the CallRetry table which means we
            // never got a CDR from IMI, so let's warn about this and still try to reschedule it for this week.
            Long id = callRetry.getId();
            LOGGER.warn(String.format("CallRetry record (id %d) for subscription %s for weekId %s was never " +
                            "deleted, which means we never received a CDR about it. Someone should look into this.", id,
                    callRetry.getSubscriptionId(), callRetry.getWeekId()));
            callRetry.setCallStage(CallStage.RETRY_1);
            callRetry.setWeekId(record.getWeekId());
            callRetryDataService.update(callRetry);
            return;
        }

        // We've already rescheduled this call, let's see if it needs to be re-rescheduled
        if ((subscription.getSubscriptionPack().retryCount() == 1) ||
                (callRetry.getCallStage() == CallStage.RETRY_LAST)) {